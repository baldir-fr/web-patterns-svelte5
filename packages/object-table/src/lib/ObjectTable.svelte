<script lang="ts">
    import ObjectTable from "./ObjectTable.svelte"

    const {data, summarize, summarizeInner} = $props()

    let typeOfData = $derived(Object.prototype.toString.call(data))
    let isArray = $derived(typeOfData === "[object Array]")

    let isString = $derived(typeOfData === "[object String]")
    let isNumber = $derived(typeOfData === "[object Number]")
    let isHttpUrl = $derived(isString && _isHttpUrl(data))
    let isPicture = $derived(isString && _isPicture(data))

    const _isHttpUrl = (url: string) => {
        try {
            const parsedUrl = new URL(url);
            return parsedUrl.protocol === "http:" || parsedUrl.protocol === "https:";
        } catch (_) {
            return false;
        }
    }
    const _isPicture = (url: string) => url.endsWith(".jpg")
    let isObject = $derived(typeOfData === "[object Object]")
    let isNullish = $derived(typeOfData === "[object Undefined]" || typeOfData === "[object Null]")
    let isEmptyObject = $derived(typeOfData === "[object Object]" && JSON.stringify(data) === "{}")


    function sameMembers(arr1, arr2) {
        const set1 = new Set(arr1);
        const set2 = new Set(arr2);
        return arr1.every(item => set2.has(item)) &&
            arr2.every(item => set1.has(item))
    }

    const _isArrayOfSameObjects = (anArray: Array<unknown>): boolean => {

        const dataType = Object.prototype.toString.call(anArray);
        if (dataType !== "[object Array]") {
            return false
        }
        if (anArray.length === 0) {
            return false
        }
        const every_array_value_is_an_object = anArray.every(value => Object.prototype.toString.call(value) === "[object Object]");
        if (!every_array_value_is_an_object) {
            return false;
        }
        const first_array_value_keys = Object.keys(anArray[0])
        const every_array_value_has_the_same_keys = anArray.every(value => sameMembers(first_array_value_keys, Object.keys(value)))
        return every_array_value_has_the_same_keys;

    }

    let isArrayOfSameObjects = $derived(_isArrayOfSameObjects(data))

    const deriveHeaders = (someData) => {
        if (someData === undefined) {
            return []
        }
        if (isObject) {
            return Object.keys(someData)
        }
        if (isArrayOfSameObjects) {
            return Object.keys(someData[0])
        }
        return []
    }

    const summarizeList = (items: string[]) => {
        const joinedItems = items.join(", ");
        const maxLength = 100;
        if (joinedItems.length <= maxLength) {
            return joinedItems
        }
        return joinedItems.slice(0, maxLength) + "â€¦";
    }

    const deriveRows = (someData) => {
        if (someData === undefined) {
            return []
        }
        if (isObject) {
            return [someData]
        }
        if (isArrayOfSameObjects) {
            console.log("deriveRows.isArrayOfSameObjects")
            return someData
        }
        return []
    }

    let headers = $derived(deriveHeaders(data))
    let rows = $derived(deriveRows(data))

    // $inspect("data : " + JSON.stringify(data))
    // $inspect("isString : " + isString + ", isObject : " + isObject + ", isArray : " + isArray + ", isArrayOfSameObjects : " + isArrayOfSameObjects + ", isEmptyObject: " + isEmptyObject + ", isNullish : " + isNullish)
    // $inspect("rows : " + JSON.stringify(rows))
    // $inspect("headers : " + JSON.stringify(headers))
    // $inspect("isArrayOfSameObjects : " + isArrayOfSameObjects)

    // const headers: Array<string> = ["a", "b", "c"]
    // const rows: Array<Record<string, unknown>> = [
    //     {
    //         "a": "foo",
    //         "b": "bar"
    //     },
    //     {
    //         "b": "bar bar",
    //         "a": "foo foo",
    //     },
    // ]


</script>

{#if isPicture}{@render picture(data)}

{:else if isHttpUrl}{@render link(data)}

{:else if isString || isNumber}{@render string(data)}

{:else if isNullish || isEmptyObject}{@render empty()}

{:else if isArray && !isArrayOfSameObjects}
    {#if summarize}
        <details>
            <summary>{summarizeList(data)}</summary>
            {@render list(data, summarizeInner)}
        </details>
    {:else}
        {@render list(data, summarizeInner)}
    {/if}
{:else if isObject || isArrayOfSameObjects}
    {#if summarize}
        <details>
            <summary>{summarizeList(headers)}</summary>
            {@render table(data, summarizeInner)}
        </details>
    {:else}
        {@render table(data, summarizeInner)}
    {/if}
{/if}

{#snippet list(data, summarizeInner)}
    <ul>
        {#each data as stringValue}
            <li>
                <ObjectTable data={stringValue} summarize={summarizeInner}/>
            </li>
        {/each}
    </ul>
{/snippet}

{#snippet table(data, summarizeInner)}
    <table class="object-table-table">
        <thead>
        <tr>
            {#each headers as header}
                <th>{header}</th>
            {/each}
        </tr>
        </thead>
        <tbody>
        {#each rows as row}
            <tr>
                {#each headers as header}
                    <td>
                        <ObjectTable data={row[header]} summarize={summarizeInner}/>
                    </td>
                {/each}
            </tr>
        {/each}
        </tbody>
    </table>
{/snippet}

{#snippet picture(data)}
    <img src={data}/>
{/snippet}

{#snippet link(data)}
    <a href={data}>{data}</a>
{/snippet}

{#snippet string(data)}
    <span>{data}</span>
{/snippet}

{#snippet empty()}<span></span>{/snippet}


<style>
    /* Global style to embrace the cascade. */
    :global {
        .object-table-table table, .object-table-table td, .object-table-table th {
            border: lightslategrey solid;
        }

        table.object-table-table {
            border-collapse: collapse;
        }

        .object-table-table td {
            min-width: 12rem;
            padding: 1rem;
            overflow-wrap: break-word;
            overflow-x: auto;
        }

        .object-table-table th {

            padding: 1rem;
        }
    }
</style>
