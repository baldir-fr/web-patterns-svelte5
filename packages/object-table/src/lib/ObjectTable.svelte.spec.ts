import {describe, expect, it} from "vitest";
import {render} from "@testing-library/svelte";
import ObjectTable from "$lib/ObjectTable.svelte";
import '@testing-library/jest-dom/vitest';

describe('ObjectTable', () => {

    describe('with empty data', () => {
        it.each([
            {data: undefined},
            {data: null},
        ])('renders no element with $data', ({data}) => {
            const {container} = render(ObjectTable, {
                props: {
                    data
                }
            })
            expect(container.textContent).toBe("")
        })
    })
    describe('with string literal', () => {
        it('renders as text', () => {
            const {getAllByText} = render(ObjectTable, {
                props: {
                    data: "some text"
                }
            })

            getAllByText("some text")
        })
        it('escapes text', () => {
            const {getByText} = render(ObjectTable, {
                props: {
                    data: `<img src="foo.png" />`
                }
            })

            const result = getByText(`<img src="foo.png" />`)

            expect(result).toMatchInlineSnapshot(`
              <span>
                &lt;img src="foo.png" /&gt;
              </span>
            `)
        })
    })
    describe('with HTTP Url', () => {
        it('renders as anchor', () => {
            const {getByText} = render(ObjectTable, {
                props: {
                    data: "https://foo.bar"
                }
            })


            expect(getByText("https://foo.bar")).toMatchInlineSnapshot(`
              <a
                href="https://foo.bar"
              >
                https://foo.bar
              </a>
            `)
        })
    })
    describe('with Picture URL', () => {
        it('renders as image', () => {
            const {getByRole} = render(ObjectTable, {
                props: {
                    data: "https://foo.bar.jpg"
                }
            })

            expect(getByRole("presentation")).toMatchInlineSnapshot(`
              <img
                alt=""
                src="https://foo.bar.jpg"
              />
            `)
        })
    })
    describe('with string array', () => {
        it('renders as an unordered list', () => {
            const {getByRole} = render(ObjectTable, {
                props: {
                    data: ["a", "b", "c"]
                }
            })

            const list = getByRole("list")
            expect(list).toMatchInlineSnapshot(`
              <ul>
                <li>
                  <!---->
                  <span>
                    a
                  </span>
                  
                  <!---->
                </li>
                <li>
                  <!---->
                  <span>
                    b
                  </span>
                  
                  <!---->
                </li>
                <li>
                  <!---->
                  <span>
                    c
                  </span>
                  
                  <!---->
                </li>
                
              </ul>
            `)
        });

        it('renders as an empty list', () => {
            const {getByRole} = render(ObjectTable, {
                props: {
                    data: []
                }
            })

            const list = getByRole("list")
            expect(list).toMatchInlineSnapshot(`
              <ul>
                
              </ul>
            `)
        });
    });
    describe('with flat object', () => {
        // tag::global-style-rendered[]
        it('renders as table with attributes as headers', () => {
            const {getByRole} = render(ObjectTable, {props: {data: {foo: "bar"}}})
            const table = getByRole("table")
            expect(table).toMatchInlineSnapshot(`
              <table
                class="object-table-table"
              >
                <thead>
                  <tr>
                    <th>
                      foo
                    </th>
                    
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td>
                      <!---->
                      <span>
                        bar
                      </span>
                      
                      <!---->
                    </td>
                    
                  </tr>
                  
                </tbody>
              </table>
            `)
        });
        // end::global-style-rendered[]
    });

    describe('with nested object', () => {
        it('renders as table with nested table in cell', () => {
            const {getAllByRole} = render(ObjectTable, {
                props: {
                    data:
                        {
                            foo: {
                                bar: "baz"
                            }
                        }
                }
            })
            const tables = getAllByRole("table")
            const outerTable = tables[0]
            expect(outerTable).toMatchInlineSnapshot(`
              <table
                class="object-table-table"
              >
                <thead>
                  <tr>
                    <th>
                      foo
                    </th>
                    
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td>
                      <!---->
                      <!---->
                      <table
                        class="object-table-table"
                      >
                        <thead>
                          <tr>
                            <th>
                              bar
                            </th>
                            
                          </tr>
                        </thead>
                        <tbody>
                          <tr>
                            <td>
                              <!---->
                              <span>
                                baz
                              </span>
                              
                              <!---->
                            </td>
                            
                          </tr>
                          
                        </tbody>
                      </table>
                      
                      
                      <!---->
                    </td>
                    
                  </tr>
                  
                </tbody>
              </table>
            `)

            const innerTable = tables[1]
            expect(innerTable).toMatchInlineSnapshot(`
              <table
                class="object-table-table"
              >
                <thead>
                  <tr>
                    <th>
                      bar
                    </th>
                    
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td>
                      <!---->
                      <span>
                        baz
                      </span>
                      
                      <!---->
                    </td>
                    
                  </tr>
                  
                </tbody>
              </table>
            `)
        });
    });
    describe('with array of objects with same keys', () => {
        it('renders as table inner objects keys as headers', () => {
            const {getAllByRole} = render(ObjectTable, {
                props: {
                    data:
                        [
                            {foo: "bar", baz: "moo"},
                            {foo: "bar 2", baz: "moo 2"},
                        ]
                }
            })
            const tables = getAllByRole("table")
            const outerTable = tables[0]
            expect(outerTable).toMatchInlineSnapshot(`
              <table
                class="object-table-table"
              >
                <thead>
                  <tr>
                    <th>
                      foo
                    </th>
                    <th>
                      baz
                    </th>
                    
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td>
                      <!---->
                      <span>
                        bar
                      </span>
                      
                      <!---->
                    </td>
                    <td>
                      <!---->
                      <span>
                        moo
                      </span>
                      
                      <!---->
                    </td>
                    
                  </tr>
                  <tr>
                    <td>
                      <!---->
                      <span>
                        bar 2
                      </span>
                      
                      <!---->
                    </td>
                    <td>
                      <!---->
                      <span>
                        moo 2
                      </span>
                      
                      <!---->
                    </td>
                    
                  </tr>
                  
                </tbody>
              </table>
            `)

            const innerTable = tables[1]
            expect(innerTable).toMatchInlineSnapshot(`undefined`)
        });
    });

    describe('with array of objects with different keys', () => {
        it('renders as table inner objects keys as headers', () => {
            const {getAllByRole} = render(ObjectTable, {
                props: {
                    data:
                        [
                            {foo: "bar"},
                            {foo: "bar 2", baz: "moo"},
                        ]
                }
            })
            const tables = getAllByRole("list")
            const outerTable = tables[0]
            expect(outerTable).toMatchInlineSnapshot(`
              <ul>
                <li>
                  <!---->
                  <!---->
                  <table
                    class="object-table-table"
                  >
                    <thead>
                      <tr>
                        <th>
                          foo
                        </th>
                        
                      </tr>
                    </thead>
                    <tbody>
                      <tr>
                        <td>
                          <!---->
                          <span>
                            bar
                          </span>
                          
                          <!---->
                        </td>
                        
                      </tr>
                      
                    </tbody>
                  </table>
                  
                  
                  <!---->
                </li>
                <li>
                  <!---->
                  <!---->
                  <table
                    class="object-table-table"
                  >
                    <thead>
                      <tr>
                        <th>
                          foo
                        </th>
                        <th>
                          baz
                        </th>
                        
                      </tr>
                    </thead>
                    <tbody>
                      <tr>
                        <td>
                          <!---->
                          <span>
                            bar 2
                          </span>
                          
                          <!---->
                        </td>
                        <td>
                          <!---->
                          <span>
                            moo
                          </span>
                          
                          <!---->
                        </td>
                        
                      </tr>
                      
                    </tbody>
                  </table>
                  
                  
                  <!---->
                </li>
                
              </ul>
            `)

            const innerTable = tables[1]
            expect(innerTable).toMatchInlineSnapshot(`undefined`)
        });
    });
});