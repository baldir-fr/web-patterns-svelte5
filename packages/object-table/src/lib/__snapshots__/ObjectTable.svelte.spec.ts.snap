// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ObjectTable > with empty data > renders no element with null > dssd 1`] = `
<div>
  <!---->
  <!---->
   
  <!---->
   
  <!---->
  
</div>
`;

exports[`ObjectTable > with empty data > renders no element with undefined > dssd 1`] = `
<div>
  <!---->
  <span />
  <!---->
   
  <!---->
   
  <!---->
  
</div>
`;
